File: mirea/cources/2026-2027/aii-july/01/Part-03-Sorting/examples/example-08.cpp
// Демонстрация stable_sort в C++

#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

struct Student {
    int score;
    string name;
};

int main() {
    // В исходном порядке студенты с одинаковым score идут друг за другом
    vector<Student> students = {
        {2, "Ann"},
        {1, "Boris"},
        {2, "Clara"},
        {1, "Diana"},
        {2, "Eva"},
    };

    // Устойчивая сортировка по score сохраняет порядок равных элементов
    stable_sort(students.begin(), students.end(), [](const Student& left, const Student& right) {
        return left.score < right.score;
    });

    // Вывод результата
    for (const Student& student : students) {
        cout << student.score << ':' << student.name << '\n';
    }
}

File: mirea/cources/2026-2027/aii-july/01/Part-03-Sorting/examples/example-08.py
# Демонстрация устойчивой сортировки в Python3

# Встроенная сортировка Python3 уже является устойчивой
students = [
    (2, "Ann"),
    (1, "Boris"),
    (2, "Clara"),
    (1, "Diana"),
    (2, "Eva"),
]

# Сортируем по первому элементу пары и сохраняем порядок равных элементов
students.sort(key=lambda item: item[0])

# Вывод результата
for score, name in students:
    print(f"{score}:{name}")